File: addons/terrabrush/Scripts/ZoneUtils.cs
using System;
using Godot;

namespace TerraBrush;

public static class ZoneUtils {
    private const string HeightmapFileName = "Heightmap_{0}_{1}.res";
    private const string SplatmapFileName = "Splatmap_{0}_{1}_{2}.res";
    private const string FoliageFileName = "Foliage_{0}_{1}_{2}.res";
    private const string ObjectFileName = "Object_{0}_{1}_{2}.res";
    private const string WaterFileName = "Water_{0}_{1}.res";
    private const string SnowFileName = "Snow_{0}_{1}.res";

    public static ImageTexture CreateLockImage(int zoneSize, Vector2I zonePosition, bool lockAll = false) {
        var image = GodotAgnostic.ImageCreateEmpty(zoneSize, zoneSize, false, Image.Format.Rf);
        if (lockAll) {
            image.Fill(Colors.White);
        }

        return GetImageTextureResource(image, string.Format(HeightmapFileName, zonePosition.X, zonePosition.Y), "");
    }

    public static ImageTexture CreateHeightmapImage(int zoneSize, float resolution, Vector2I zonePosition, string dataPath) {
        var imageSize = GetImageSizeForResolution(zoneSize, resolution);
        var image = GodotAgnostic.ImageCreateEmpty(imageSize, imageSize, false, Image.Format.Rgf);
        return GetImageTextureResource(image, string.Format(HeightmapFileName, zonePosition.X, zonePosition.Y), dataPath);
    }

    public static ImageTexture CreateSplatmapImage(int zoneSize, Vector2I zonePosition, int splatmapIndex, string dataPath) {
        var splatmapImage = GodotAgnostic.ImageCreateEmpty(zoneSize, zoneSize, false, Image.Format.Rgba8);

        if (splatmapIndex == 0) {
            splatmapImage.Fill(new Color(1, 0, 0, 0));
        } else {
            splatmapImage.Fill(new Color(0, 0, 0, 0));
        }

        return GetImageTextureResource(splatmapImage, string.Format(SplatmapFileName, zonePosition.X, zonePosition.Y, splatmapIndex), dataPath);
    }

    public static ImageTexture CreateFoliageImage(int zoneSize, Vector2I zonePosition, int foliageIndex, string dataPath) {
        var image = GodotAgnostic.ImageCreateEmpty(zoneSize, zoneSize, false, Image.Format.Rgba8);
        return GetImageTextureResource(image, string.Format(FoliageFileName, zonePosition.X, zonePosition.Y, foliageIndex), dataPath);
    }

    public static ImageTexture CreateObjectImage(int zoneSize, Vector2I zonePosition, int objectIndex, string dataPath) {
        var image = GodotAgnostic.ImageCreateEmpty(zoneSize, zoneSize, false, Image.Format.Rgba8);
        return GetImageTextureResource(image, string.Format(ObjectFileName, zonePosition.X, zonePosition.Y, objectIndex), dataPath);
    }

    public static ImageTexture CreateWaterImage(int zoneSize, float resolution, Vector2I zonePosition, string dataPath) {
        var imageSize = GetImageSizeForResolution(zoneSize, resolution);
        var waterImage = GodotAgnostic.ImageCreateEmpty(imageSize, imageSize, false, Image.Format.Rgba8);
        waterImage.Fill(new Color(0, 0.5f, 0.5f, 1));

        return GetImageTextureResource(waterImage, string.Format(WaterFileName, zonePosition.X, zonePosition.Y), dataPath);
    }

    public static ImageTexture CreateSnowImage(int zoneSize, float resolution, Vector2I zonePosition, string dataPath) {
        var imageSize = GetImageSizeForResolution(zoneSize, resolution);
        var snowImage = GodotAgnostic.ImageCreateEmpty(imageSize, imageSize, false, Image.Format.Rgba8);
        return GetImageTextureResource(snowImage, string.Format(SnowFileName, zonePosition.X, zonePosition.Y), dataPath);
    }

    private static ImageTexture GetImageTextureResource(Image image, string filePath, string dataPath) {
        ImageTexture imageTexture;
        if (string.IsNullOrWhiteSpace(dataPath)) {
            imageTexture = ImageTexture.CreateFromImage(image);
        } else {
            var resourcePath = Utils.PathCombineForwardSlash(dataPath, filePath);
            if (Godot.FileAccess.FileExists(resourcePath)) {
                imageTexture = ResourceLoader.Load<ImageTexture>(resourcePath);
                imageTexture.SetImage(image);
            } else {
                imageTexture = ImageTexture.CreateFromImage(image);
                ResourceSaver.Save(imageTexture, resourcePath);
                imageTexture = ResourceLoader.Load<ImageTexture>(resourcePath);
            }
        }

        return imageTexture;
    }

    public static ZoneInfo GetPixelToZoneInfo(float x, float y, int zonesSize, int resolution) {
        if (zonesSize % 2 == 0) {
            x -= 0.5f;
            y -= 0.5f;
        }

        var zoneXPosition = Mathf.FloorToInt(x / (zonesSize - 1));
        var zoneYPosition = Mathf.FloorToInt(y / (zonesSize - 1));
        var zonePosition = new Vector2I(zoneXPosition, zoneYPosition);

        var zoneBrushXPosition = Mathf.RoundToInt(((x / (zonesSize - 1)) - zoneXPosition) * (zonesSize - 1));
        var zoneBrushYPosition = Mathf.RoundToInt(((y / (zonesSize - 1)) - zoneYPosition) * (zonesSize - 1));

        var resolutionZoneBrushXPosition = zoneBrushXPosition;
        var resolutionZoneBrushYPosition = zoneBrushYPosition;

        if (resolution != 1) {
            var imageSize = GetImageSizeForResolution(zonesSize, resolution);

            resolutionZoneBrushXPosition = Mathf.RoundToInt(Mathf.Remap(resolutionZoneBrushXPosition, 0, zonesSize - 1, 0, imageSize - 1));
            resolutionZoneBrushYPosition = Mathf.RoundToInt(Mathf.Remap(resolutionZoneBrushYPosition, 0, zonesSize - 1, 0, imageSize - 1));
        }

        // This is just a unique key that combines the x and y, perfect to keep the zone info in cache.
        var zoneKey = (zonePosition.X << 8) + zonePosition.Y;

        return new ZoneInfo() {
            ZoneKey = zoneKey,
            ZonePosition = zonePosition,
            ImagePosition = new Vector2I(resolutionZoneBrushXPosition, resolutionZoneBrushYPosition)
        };
    }

    public static ZoneInfo GetZoneInfoFromZoneOffset(ZoneInfo startingZone, Vector2I offset, int zonesSize, int resolution) {
        var resolutionSize = GetImageSizeForResolution(zonesSize, resolution);

        var pixelPosition = new Vector2(startingZone.ImagePosition.X + offset.X, startingZone.ImagePosition.Y + offset.Y);
        var zoneXPosition = Mathf.FloorToInt(pixelPosition.X / resolutionSize);
        var zoneYPosition = Mathf.FloorToInt(pixelPosition.Y / resolutionSize);

        var zoneBrushXPosition = Mathf.RoundToInt(((pixelPosition.X / resolutionSize) - zoneXPosition) * resolutionSize);
        var zoneBrushYPosition = Mathf.RoundToInt(((pixelPosition.Y / resolutionSize) - zoneYPosition) * resolutionSize);

        var resolutionZoneBrushXPosition = zoneBrushXPosition;
        var resolutionZoneBrushYPosition = zoneBrushYPosition;

        // This is just a unique key that combines the x and y, perfect to keep the zone info in cache.
        var absoluteZonePosition = new Vector2I(startingZone.ZonePosition.X + zoneXPosition, startingZone.ZonePosition.Y + zoneYPosition);
        var zoneKey = (absoluteZonePosition.X << 8) + absoluteZonePosition.Y;

        return new ZoneInfo() {
            ZoneKey = zoneKey,
            ZonePosition = absoluteZonePosition,
            ImagePosition = new Vector2I(resolutionZoneBrushXPosition, resolutionZoneBrushYPosition)
        };
    }

    public static int GetImageSizeForResolution(int zoneSize, float resolution) {
        return Mathf.CeilToInt(zoneSize / resolution);
    }
}

public class ZoneInfo {
    public int ZoneKey { get;set; }
    public Vector2I ZonePosition { get;set; }
    public Vector2I ImagePosition { get;set; }
}

File: src/misc/zone_utils.cpp
#include "zone_utils.h"
#include "utils.h"

#include <godot_cpp/classes/image.hpp>
#include <godot_cpp/classes/file_access.hpp>
#include <godot_cpp/classes/resource_loader.hpp>
#include <godot_cpp/classes/resource_saver.hpp>

using namespace godot;

Ref<Image> ZoneUtils::createLockImage(int zoneSize, Vector2i zonePosition, bool lockAll) {
    Ref<Image> image = Image::create_empty(zoneSize, zoneSize, false, Image::Format::FORMAT_RF);
    if (lockAll) {
        image->fill(Color::named("WHITE"));
    }

    return getImageResource(image, String(HeightmapFileName).format(Array::make(zonePosition.x, zonePosition.y)), "");
}

Ref<Image> ZoneUtils::createHeightmapImage(int zoneSize, float resolution, Vector2i zonePosition, String dataPath) {
    int imageSize = getImageSizeForResolution(zoneSize, resolution);
    Ref<Image> image = Image::create_empty(imageSize, imageSize, false, Image::Format::FORMAT_RGF);
    return getImageResource(image, String(HeightmapFileName).format(Array::make(zonePosition.x, zonePosition.y)), dataPath);
}

Ref<Image> ZoneUtils::createSplatmapImage(int zoneSize, Vector2i zonePosition, int splatmapIndex, String dataPath) {
    Ref<Image> splatmapImage = Image::create_empty(zoneSize, zoneSize, false, Image::Format::FORMAT_RGBA8);

    if (splatmapIndex == 0) {
        splatmapImage->fill(Color(1, 0, 0, 0));
    } else {
        splatmapImage->fill(Color(0, 0, 0, 0));
    }

    return getImageResource(splatmapImage, String(SplatmapFileName).format(Array::make(zonePosition.x, zonePosition.y, splatmapIndex)), dataPath);
}

Ref<Image> ZoneUtils::createFoliageImage(int zoneSize, Vector2i zonePosition, int foliageIndex, String dataPath) {
    Ref<Image> image = Image::create_empty(zoneSize, zoneSize, false, Image::Format::FORMAT_RGBA8);
    return getImageResource(image, String(FoliageFileName).format(Array::make(zonePosition.x, zonePosition.y, foliageIndex)), dataPath);
}

Ref<Image> ZoneUtils::createObjectImage(int zoneSize, Vector2i zonePosition, int objectIndex, String dataPath) {
    Ref<Image> image = Image::create_empty(zoneSize, zoneSize, false, Image::Format::FORMAT_RGBA8);
    return getImageResource(image, String(ObjectFileName).format(Array::make(zonePosition.x, zonePosition.y, objectIndex)), dataPath);
}

Ref<Image> ZoneUtils::createWaterImage(int zoneSize, float resolution, Vector2i zonePosition, String dataPath) {
    int imageSize = getImageSizeForResolution(zoneSize, resolution);
    Ref<Image> waterImage = Image::create_empty(imageSize, imageSize, false, Image::Format::FORMAT_RGBA8);
    waterImage->fill(Color(0, 0.5f, 0.5f, 1));

    return getImageResource(waterImage, String(WaterFileName).format(Array::make(zonePosition.x, zonePosition.y)), dataPath);
}

Ref<Image> ZoneUtils::createSnowImage(int zoneSize, float resolution, Vector2i zonePosition, String dataPath) {
    int imageSize = getImageSizeForResolution(zoneSize, resolution);
    Ref<Image> snowImage = Image::create_empty(imageSize, imageSize, false, Image::Format::FORMAT_RGBA8);
    return getImageResource(snowImage, String(SnowFileName).format(Array::make(zonePosition.x, zonePosition.y)), dataPath);
}

Ref<Image> ZoneUtils::createMetaInfoImage(int zoneSize, float resolution, Vector2i zonePosition, String dataPath) {
    int imageSize = getImageSizeForResolution(zoneSize, resolution);
    Ref<Image> metaInfoImage = Image::create_empty(imageSize, imageSize, false, Image::Format::FORMAT_RF);
    metaInfoImage->fill(Color(-1, 0, 0, 0));

    return getImageResource(metaInfoImage, String(MetaInfoFileName).format(Array::make(zonePosition.x, zonePosition.y)), dataPath);
}

Ref<Image> ZoneUtils::getImageResource(Ref<Image> image, String filePath, String dataPath) {
    if (!dataPath.is_empty()) {
        String resourcePath = Utils::pathCombineForwardSlash(dataPath, filePath);
        if (!FileAccess::file_exists(resourcePath)) {
            ResourceSaver::get_singleton()->save(image, resourcePath);
            image = ResourceLoader::get_singleton()->load(resourcePath);
        }
    }

    return image;
}

ZoneInfo ZoneUtils::getPixelToZoneInfo(float x, float y, int zonesSize, int resolution) {
    if (zonesSize % 2 == 0) {
        x -= 0.5f;
        y -= 0.5f;
    }

    int zoneXPosition = (int) Math::floor(x / (zonesSize - 1));
    int zoneYPosition = (int) Math::floor(y / (zonesSize - 1));
    Vector2i zonePosition = Vector2i(zoneXPosition, zoneYPosition);

    int zoneBrushXPosition = (int) Math::round(((x / (zonesSize - 1)) - zoneXPosition) * (zonesSize - 1));
    int zoneBrushYPosition = (int) Math::round(((y / (zonesSize - 1)) - zoneYPosition) * (zonesSize - 1));

    int resolutionZoneBrushXPosition = zoneBrushXPosition;
    int resolutionZoneBrushYPosition = zoneBrushYPosition;

    if (resolution != 1) {
        int imageSize = getImageSizeForResolution(zonesSize, resolution);

        resolutionZoneBrushXPosition = (int) Math::round(Math::remap((float) resolutionZoneBrushXPosition, 0, zonesSize - 1, 0, imageSize - 1));
        resolutionZoneBrushYPosition = (int) Math::round(Math::remap((float) resolutionZoneBrushYPosition, 0, zonesSize - 1, 0, imageSize - 1));
    }

    // This is just a unique key that combines the x and y, perfect to keep the zone info in cache.
    int zoneKey = (zonePosition.x << 8) + zonePosition.y;

    return ZoneInfo(zoneKey, zonePosition, Vector2i(resolutionZoneBrushXPosition, resolutionZoneBrushYPosition));
}

ZoneInfo ZoneUtils::getZoneInfoFromZoneOffset(ZoneInfo &startingZone, Vector2i offset, int zonesSize, int resolution) {
    int resolutionSize = getImageSizeForResolution(zonesSize, resolution);
    Vector2 pixelPosition = Vector2(startingZone.imagePosition.x + offset.x, startingZone.imagePosition.y + offset.y);
    int zoneXPosition = (int) Math::floor(pixelPosition.x / resolutionSize);
    int zoneYPosition = (int) Math::floor(pixelPosition.y / resolutionSize);

    int zoneBrushXPosition = (int) Math::round(((pixelPosition.x / resolutionSize) - zoneXPosition) * resolutionSize);
    int zoneBrushYPosition = (int) Math::round(((pixelPosition.y / resolutionSize) - zoneYPosition) * resolutionSize);

    int resolutionZoneBrushXPosition = zoneBrushXPosition;
    int resolutionZoneBrushYPosition = zoneBrushYPosition;

    // This is just a unique key that combines the x and y, perfect to keep the zone info in cache.
    Vector2i absoluteZonePosition = Vector2i(startingZone.zonePosition.x + zoneXPosition, startingZone.zonePosition.y + zoneYPosition);
    int zoneKey = (absoluteZonePosition.x << 8) + absoluteZonePosition.y;

    return ZoneInfo(zoneKey, absoluteZonePosition, Vector2i(resolutionZoneBrushXPosition, resolutionZoneBrushYPosition));
}

int ZoneUtils::getImageSizeForResolution(int zoneSize, float resolution) {
    return (int) Math::ceil(zoneSize / resolution);
}
